bundlewrap/utils/dicts.py

```diff
@@ -183,6 +191,29 @@ def default(self, obj):
             return JSONEncoder.default(self, obj)
 
 
+def freeze_object(obj):
+    """
+    Returns a read-only version of the given object (if possible).
+    """
+    if isinstance(obj, dict):
+        keys = set(obj.keys())
+        for k in keys:
+            obj[k] = freeze_object(obj[k])
+        return MappingProxyType(obj)
+    elif isinstance(obj, (list, tuple)):
+        result = []
+        for i in obj:
+            result.append(freeze_object(i))
+        return tuple(result)
+    elif isinstance(obj, set):
+        result = set()
+        for i in obj:
+            result.add(freeze_object(i))
+        return frozenset(obj)
+    else:
+        return obj
```

Comment:
we could use this opportunity to check for other invalid types (see https://docs.bundlewrap.org/repo/nodes.py/#metadata)